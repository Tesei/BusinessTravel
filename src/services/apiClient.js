// src/apiClient.js
import axios from 'axios'
import { useNotificationStore } from '@/store/notification'
import { useUserStore } from '@/store/user'

// Базовая настройка клиента axios
const apiClient = axios.create({
    baseURL: import.meta.env.VITE_AUTH_BTUI_SERVICE, // базовый URL
    timeout: 10000,
    headers: {
        'Content-Type': 'application/json',
    },
})

// Перехватчик запроса для токена
apiClient.interceptors.request.use(
    (config) => {
        const typeToken = useUserStore().incomingTokens.type
        const accessToken = useUserStore().incomingTokens.accessToken
        const autToken = typeToken && accessToken ? typeToken + ' ' + accessToken : ''
        if (autToken) config.headers.Authorization = autToken
        return config
    },
    (error) => {
        return Promise.reject(error)
    },
)

// Перехватчик ответа для отслеживания ошибок
apiClient.interceptors.response.use(
    (response) => {
        return response
    },
    (error) => {
        // Обработка ошибок
        if (error.response) {
            // Сервер вернул ответ с ошибкой
            console.error('Error:', error.response.data.message || 'Ошибка сервера')
            useNotificationStore().setError(error)
        } else {
            console.error('Network Error:', error.message || 'Ошибка сети')
        }
        return Promise.reject(error)
    },
)

export default apiClient
