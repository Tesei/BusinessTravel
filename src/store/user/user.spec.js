import { describe, it, expect, beforeEach, vi, afterEach } from 'vitest'
import { setActivePinia, createPinia } from 'pinia'
import { useUserStore } from './index'
import apiClient from '@/services/apiClient'

import flushPromises from 'flush-promises'
import waitForExpect from 'wait-for-expect'

describe('useUserStore', () => {
    beforeEach(() => {
        // создаем новый экземпляр pinia и делаем его активным,
        // чтобы он автоматически подхватывается любым вызовом useStore()
        // без необходимости его передачи: `useStore(pinia)`
        setActivePinia(createPinia())
        // vi.mock('@/services/apiClient', () => ({
        //     __esModule: true,
        //     post: vi.fn(),
        // }))
    })

    afterEach(() => {
        // очищаем моки
        vi.resetAllMocks()
    })

    it('imports without error', () => {
        expect(useUserStore).toBeDefined()
        expect(useUserStore).toMatchSnapshot()
    })
    it('наличие всех начальных переменных state', () => {
        expect(useUserStore().isLoggedIn).toBeDefined()
        expect(useUserStore().changePasswordAlerts).toBeDefined()
        expect(useUserStore().isUserFound).toBeDefined()
        expect(useUserStore().directions).toBeDefined()
        expect(useUserStore().user).toBeDefined()
        expect(useUserStore().incomingTokens).toBeDefined()
        expect(useUserStore().counterpartiesFromID).toBeDefined()
    })

    // it('defaultGoods array has items', () => {
    //     const goodsStore = useUserStore()
    //     expect(goodsStore.defaultGoods).toHaveLength(8)
    // })

    it('запрос токена, проверка fetchLogIn', async () => {
        // ! Arrange
        const userStore = useUserStore()
        // eslint-disable-next-line no-import-assign
        apiClient = vi.fn().mockResolvedValue({
            status: 200,
            data: { 
                accessToken: 'eyJhbGciOiJIUzI1NiIsI',
                refreshToken: 'NjM4NzU3MjQxODgyMzU',
                refreshTokenExpiresIn: '2025-02-21T10:43:08.2350858Z',
            },
        })        
        // ! Act
        await userStore.fetchLogIn({
            login: 'some@mail.ru',
            password: '1231232532',
        })
        // ! Assert
        await flushPromises()         
        await waitForExpect(() => {
            expect(userStore.incomingTokens.accessToken).toBe('eyJhbGciOiJIUzI1NiIsI')
            expect(userStore.isLoggedIn).toBe(true)
        })
    })
    it('проверка выхода, logout', async () => {
        const userStore = useUserStore()
        // eslint-disable-next-line no-import-assign
        apiClient = vi.fn().mockResolvedValue({
            status: 200,
            data: {},
        })        

        await userStore.logout()
        await flushPromises()         
        await waitForExpect(() => {
            expect(userStore.isLoggedIn).toBe(false)
        })
    })
})