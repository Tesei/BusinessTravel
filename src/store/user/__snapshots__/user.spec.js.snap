// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useUserStore > imports without error 1`] = `[Function]`;
